[mono] Fix vector class retrieval and type checks for binary operand APIs

matouskozak opened the pull request:
## Bug description
On Arm64, currently `Vector[64/128].Multiply(scalar, vector)` fails because we used the first argument to retrieve the vector class which in this case fails because it's not a vector.

Class was previously retrieved as a class of the first argument: https://github.com/dotnet/runtime/blob/88f9aba91e11d1695ebe9ab572736ee62ac7ad61/src/mono/mono/mini/simd-intrinsics.c#L1995 
which in this case is a scalar value not a vector. This caused a crash when we verify `is_element_type_primitive`
https://github.com/dotnet/runtime/blob/2358c592c2efb5b9e5fe48d3332f105c1b2fff0d/src/mono/mono/mini/simd-intrinsics.c#L2068-L2069
because it expects vector type.

This issue wasn't identified before because we are missing FullAOT-arm64 runtime tests coverage, and this scenario is only tested in https://github.com/dotnet/runtime/blob/main/src/tests/JIT/Regression/JitBlue/Runtime_93876/Runtime_93876.cs


## Fix
For binary operations, if the first operand is not a vector, we extract the vector type from the second operand. We also extend the fallback condition to both operands.

Additionally, full refactor of `emit_simd_ins_for_binary_op` to split the function by the OP code rather than the type of the operands.

matouskozak committed:
fix SIMD class retrievel to be vector class

matouskozak commented:
/azp run runtime-extra-platforms, runtime-llvm

jkurdek commented:
I'm wondering do we have any detection of intensified methods which do not take the correct code path. How did we find this issue, perf triage?

matouskozak commented:
> I'm wondering do we have any detection of intensified methods which do not take the correct code path. 

I'm not aware of any code analysis tool that would shows us this information. I suppose we could profile the AOT compilation to verify that we're hitting the `simd-intrinsics.c` codepaths.

> How did we find this issue, perf triage?

I found this issue while reviving the fullAOT-llvm arm64, there is one runtime test case (`Runtime_93876`) that tests this scenario. I'm not sure why we don't test this as part of libraries tests for vectors from `System.Runtime.Intrinsics`, but we do test it for vectors from `System.Numerics`. I suppose that the codegen on CoreCLR side is similar for both `Runtime.Intrinsics` and `Numerics` vectors and it was decided that testing one is enough. Do you know @tannergooding?

tannergooding commented:
We're supposed to have tests covering all the overloads, but some of the APIs or overloads may have been missed given the sheer scope of things, general timing of when various APIs were added, or may have only partial coverage.

I believe in the case of multiply here, there is partial coverage and its possible that since Mono doesn't have a "unified" implementation and takes different paths for different types and different namespaces that may have caused it to be missed in this case.

matouskozak committed:
Merge branch 'main' into fix-multiply-scalar-vector

matouskozak commented:
/azp run runtime-llvm

matouskozak committed:
retrieve vector type from second arg if necessary

matouskozak committed:
ZZMerge branch 'fix-multiply-scalar-vector' of github.com:matouskozak/runtime into fix-multiply-scalar-vector

matouskozak commented:
/azp run runtime-llvm

matouskozak commented:
/azp run runtime-extra-platforms

matouskozak committed:
Merge branch 'main' into fix-multiply-scalar-vector

tannergooding commented on src/mono/mono/mini/simd-intrinsics.c:
```diff
@@ -396,11 +396,11 @@ emit_simd_ins_for_binary_op (MonoCompile *cfg, MonoClass *klass, MonoMethodSigna
 				const char *class_name = m_class_get_name (klass);
 				if (strcmp ("Quaternion", class_name) && strcmp ("Plane", class_name)) {
-					if (!type_is_simd_vector (fsig->params [1]))
+					if (!type_is_simd_vector (fsig->params [1])) // vector * scalar
```
nit: The checks used here and elsewhere aren't really consistent

Some places are only checking the second parameter is not SIMD, others are checking it is explicitly a primitive. Likewise other checks are validating that both parameters are vector, etc.

It would be nice to be consistent here and do the right checks in all scenarios. That is, we should have at least one vector type and for the cases that accept scalars, the scalar should be a supported primitive type (not all primitive types are supported).

The intrinsics that support scalars in either position (multiply) or only in one position (divide for the second parameter) should likely be explicitly separate as well. There's some code paths, like on L2067 below, where the handling for op1 to be scalar is also being done for MinNative, Subtract, and other intrinsics where it should never be valid.

matouskozak replied:
That's a good point. I refactored both `emit_simd_ins_for_binary_op` and the callsite in `emit_sri_vector` to check for scalars and vectors for respective APIs. Let me know what you think about the changes?

matouskozak committed:
refactor `emit_simd_ins_for_binary_op``
Change the function to be split by the OP code rather than the type of the operands.
Add type checks to the callsite to ensure that the operands are of the correct type.

matouskozak committed:
Merge branch 'fix-multiply-scalar-vector' of github.com:matouskozak/runtime into fix-multiply-scalar-vector

matouskozak committed:
add missing break in switch statement

matouskozak commented:
/azp run runtime-llvm

matouskozak commented:
/ba-g WASM-debugger tests are known and hard to catch with known issue. They will be migrated from runtime soon, @thaystg @ilonatommy
